back/engines/commercial/analytics/app/controllers/analytics/web_api/v1/analytics_controller.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'query'
+
+module Analytics
+  module WebApi::V1
+    class AnalyticsController < ::ApplicationController
+      def create
+        authorize :analytics, policy_class: AnalyticsPolicy
+
+        results, errors, response_status = if params[:query].instance_of?(Array)
+          handle_multiple(params[:query])
+        else
+          handle_single(params[:query])
+        end
+
+        if [{}, []].include? errors
+          render json: { 'data' => results }
+        else
+          render json: { 'messages' => errors }, status: response_status
+        end
+      end
+
+      private
+
+      def handle_single(json_query)
+        query = Query.new(json_query)
+        query.validate
+
+        query.run if query.valid
+
+        results = query.results
+        errors = query.error_messages
+        response_status = query.response_status
+
+        [results, errors, response_status]
+      end
+
+      def handle_multiple(json_queries)
+        results = []
+        errors = {}
+        statuses = []
+        queries = []
+
+        json_queries.each_with_index do |json_query, index|
+          query = Query.new(json_query)
+          queries.push(query)
+          query.validate
+          next if query.valid
+
+          errors[index] = query.error_messages
+          statuses.push(query.response_status)
+        end
+
+        if errors == {}
+          queries.each_with_index do |query, index|
+            query.run
+            if query.failed
+              errors[index] = query.error_messages
+              statuses.push(500)
+            else
+              results.push(query.results)
+            end
+          end
+        end
+
+        [results, errors, statuses.max]
+      end
```

Comment:
the status is only being used on errors, im not very satisfied with statuses in this case but the idea is that when you have any error you get the worst status as the request status, in this case they could be 400, 422 or 500. but its not great since 400 could be worst than 422.. any suggestions?

about using `handle_single`, i didn't wan't to return any results until im sure all queries are valid.
so after making sure all are valid then i run the queries but then you could have some errors as well if some query run fails.

Reply:
To me, the choice of which status code to use is more the responsibility of the controller than the query services. The services should of course return errors, but should not be HTTP-aware. But anyway, we still need to pick a single code (be it in the controller or the service). I refer you to my previous comment to pick between 400 and 500. Now, if there is a mix of 422 (or 4**) and 400, I'd go with 400 which is more generic.

> about using handle_single, i didn't wan't to return any results until im sure all queries are valid.
so after making sure all are valid then i run the queries but then you could have some errors as well if some query run fails.

Then,  you can reverse the problem and reduce code duplication by implementing `handle_single` using `handle_multiple`. Something like (not tested):
```ruby
def handle_single(json_query)
  handle_multiple([json_query])
end
```

Reply:
this is clever, thanks, it allowed me to wrap the `json_query` into an array and just use `handle_multiple`